[in progress] Fetch results

ptone opened the pull request:
This branch is an approach to addressing:

```
Better way to record the results of a success/failed fetch-data
```

https://github.com/cartwheelweb/packaginator/issues#issue/12

This pull in NOT ready for merge, but I'm submitting it for initial review of implementation approach.

TODO:
- template changes to package
- view listing failed packages
- testing with mocks

ptone committed:
added fields to store fetch data results

ptone committed:
added fields to store fetch data results

ptone committed:
imports sort

ptone committed:
basic metadata fetch result recording

ptone committed:
basic metadata fetch result recording

ptone committed:
whitespace cleanup

ptone committed:
removed unused imports

pydanny commented:
I like this a lot! Some things that make my day:
1. Better import organization.
2. Handling failed meta-data pulls
3. Handling the pypi failures better

Keep going!

ptone commented:
if you think the 3 basic model fields will suffice, you may want to cherry pick that in, just because managing distributed migration ordering can be a bitch if you make other changes to models before this is fully merged.  They should sit there benignly until the rest of this is ready to merge in.

(https://github.com/ptone/packaginator/commit/3cfed7f31d22a0fb76687ea9550cb9c2a704f24d)

pydanny commented:
Absolutely! Great minds think alike!

In fact, Audrey and Nate Aune are working database serialization so I'm not accepting model changes until they are done.

ptone committed:
Merge branch 'master' into fetch-results

* master: (32 commits)
  Adding Aaron to contribors
  Grid view tests. Added several grid view tests; added JSON test data to confirm 404 on element view.
  Fixing classpath problem
  preloading the progress image, fixes render
  Default launchpad support is false
  Fixed a small settings issue but we now can disable launchpad without a problem
  Fixed a small settings issue but we now can disable launchpad without a problem
  added feedback during package update
  Tweak to work
  Added pagination for the profiles page
  Template tweaks to not use blocktrans.
  Added instructions/clarification to the docs.
  Fixing Chris Adams local_settings checkin.
  Add Chris Adams to contributors
  Fixed minor issue in bitbucket API handler
  Removed test_pulls from test suite, fixed minor bug in bitbucket repo handler
  Merge fun
  Handle crazy long licenses better
  Removed Sourceforge tests from __init__.py until API bug is fixed
  Edited the text. Tightened up wording.
  ...

ptone committed:
changed template_object_name for package detail

cherry-picked from template_object_rename branch

ptone committed:
add display of last metadata update

This really shouldn't be so grouped with the usage
data, but then really the whole layout of this page
probably needs a bit of a fresh approach

ptone committed:
Don't fail the update view on error in fetching metadata

add an error warning, and display the error
the failure is also recorded on the model and
displayed on package detail page

ptone commented:
This branch still needs tests, but is ready for further review if anyone is willing.

to see how it looks:
1. make a copy of your dev.db if you want to preserve the absence of the new fields
2. pull this branch into a local branch
3. run the migration
4. runserver

a quicky failure test can be simulated by disconnecting from internet (ie power off wifi) and then click the fetch latest button

the layout of package detail page needs some love, this doesn't attempt to address that

ptone committed:
reverting models file to sort out whitespace and import issues

ptone committed:
Merge branch 'master' into fetch-results

* master:
  Some docs stuff
  Add slides link
  lessons learned
  more lessons learned
  yeah
  Better organization of imports to follow pep-8
  remove extra tests file in feeds
  More docs cleanup. I'm crazy
  title fix
  expanding the faq
  Many more docs tweaks
  Adding contributing instructions
  troubleshooting things
  Search and docs changes
  Launchpad documentation additions
  repos now set in the settings
  Test for ajax_grid_list view function.
  changed template_object_name for package detail
  Removed single quote in JS that caused errors; added preventDefault().
  More test coverage for grid views.

Conflicts:
	templates/package/package.html

ptone committed:
removed duplicate manager assignment introduced after merge

ptone committed:
added test of metadata store

ptone committed:
some import additions and sorting

ptone committed:
Merge branch 'master' into fetch-results

* master:
  Removing some late night insanity
  Replaced 'import *' with class calls.
  Renamed 'package' to 'grid_package' in view as it risks clashing with other variable.
  add_gridpackage partial test coverage.
  Another view function docstring.
  View function docstring.

ptone commented:
ok - I believe this is ready to be pulled in, subject to review and comment

pydanny commented on apps/package/models.py:
```diff
@@ -128,7 +139,16 @@ class Package(BaseModel):
+            except ProtocolError, e:
+                self.fail_fetch_metadata("Error fetching pypi releases: %s" % e)
+                raise
+            except:
```
Generic exceptions are the DEVIL! ;)

ptone replied:
however in this case we are not "absorbing" any exception (the evil part) but just noting that something went wrong and immediately reraising the exact same exception.  The calling code is the one that is responsible for dealing with the exception, in this context all we want to do is "log" it.  It is generic because if an exception is raised that we didn't predict, that failure goes uncaptured.  That doesn't mean that other more specific exception handling can't go above it and improve the content of the stored message - but in this case I don't think there is evil.

ptone replied:
see also:

http://docs.python.org/tutorial/errors.html#handling-exceptions (demonstrates this exact pattern if used properly)
http://blog.ianbicking.org/good-catch-all-exceptions.html (says its OK to do if you are re-raising)

sorry if I'm seeming defensive, but your comment didn't give any suggestion on a better way, and I think this is a valid approach and want to resolve it to your satisfaction.

pydanny replied:
Nope, not defensive at all. Sometimes you have to just catch the exception and move on. Your submitted code doesn't give enough to go on. What I think I would do is:

```
try:
    DO X
except SpecificError:
    DO Y
except Exception, e:
   self.fail_fetch_metadata(str(e))
```

ptone commented:
I'm going through all the pulls and trying determine whether they are all "ready for checkin", let me know if I've addressed your concerns about the generic exception stuff - other than going through all the potential code in the call stack and looking for any possible specific exception subclasses, I'm not sure what the alternative is.  If one is missed, then the error goes unrecorded.

Also Jacob wanted this in before redoing parts of it for celery

ptone committed:
try to capture better error information on pypi error

ptone committed:
provide more information on failure of repo.fetch_metadata

ptone commented:
okeedoke, - hopefully this provides enough for the package maintainer and site admins to figure out what went wrong with metadata fetch

pydanny commented:
This will be implemented after the PyPI/Repo refactor. That refactor should not be combined with this because it will increase the complexity of the task.

pydanny commented:
REFERENCE

upcoming changes deprecate this too much for a direct merge. Keeping as reference for upcoming changes
